cmd/docker/docker.go

```diff
@@ -242,40 +248,58 @@ func tryPluginRun(dockerCli command.Cli, cmd *cobra.Command, subcommand string,
 
 	// Background signal handling logic: block on the signals channel, and
 	// notify the plugin via the PluginServer (or signal) as appropriate.
-	const exitLimit = 3
-	signals := make(chan os.Signal, exitLimit)
-	signal.Notify(signals, platformsignals.TerminationSignals...)
+	const exitLimit = 2
+
+	tryTerminatePlugin := func(force bool) {
+		// If stdin is a TTY, the kernel will forward
+		// signals to the subprocess because the shared
+		// pgid makes the TTY a controlling terminal.
+		//
+		// The plugin should have it's own copy of this
+		// termination logic, and exit after 3 retries
+		// on it's own.
+		if dockerCli.Out().IsTerminal() {
+			return
+		}
+
+		// Terminate the plugin server, which will
+		// close all connections with plugin
+		// subprocesses, and signal them to exit.
+		//
+		// Repeated invocations will result in EINVAL,
+		// or EBADF; but that is fine for our purposes.
+		_ = srv.Close()
+
+		// If we're still running after 3 interruptions
+		// (SIGINT/SIGTERM), send a SIGKILL to the plugin as a
+		// final attempt to terminate, and exit.
+		if force {
+			_, _ = fmt.Fprint(dockerCli.Err(), "got 3 SIGTERM/SIGINTs, forcefully exiting\n")
```

Comment:
very minor, but should we maybe print this termination msg outside of this func, when we do the actual check against the exitLimit?

Reply:
Yeah maybe the text doesn't quite make sense in the context of this function, but we essentially check the limit here https://github.com/docker/cli/blob/fdaad209907f8b5187697d07c452850869bc05ba/cmd/docker/docker.go#L295-L301